spec.md

```diff
@@ -72,41 +72,70 @@ We also use the following operators and functions:
 The primary result of this specification is to define a family of
 functions:
 
-$\operatorname{SPLIT}_C \in V_8 \rightarrow V_v$
+$\operatorname{SPLIT}_{T, C} \in V_8 \rightarrow V_v$
 
-...which is parameterized by a configuration $C$, consisting of:
+...which is parameterized by an integer $T \in U_{32}$, called the
+"threshold", and a configuration $C$, consisting of:
 
 - $S_{\text{min}} \in U_{32}$, the minimum split size
 - $S_{\text{max}} \in U_{32}$, the maximum split size
 - $H \in V_8 \rightarrow U_{32}$, the hash function
 - $W \in U_{32}$, the window size
-- $T \in U_{32}$, the threshold
 
 The configuration must satisfy $S_{\text{max}} \ge S_{\text{min}} \ge W > 0$.
 
 ## Definitions
 
-The "split index" $I(X)$, is either the smallest integer $i$ satisfying:
+The "split index" $I_{T, C}(X)$ is either the smallest integer $i$ satisfying:
 
 - $i < |X|$ and
 - $S_{\text{max}} \ge i \ge S_{\text{min}}$ and
 - $H(\langle X_{i-W+1}, \dots, X_i \rangle) \mod 2^T = 0$
 
 ...or $\min(|X| - 1, S_{\text{max}})$, if no such $i$ exists.
 
-We define $\operatorname{SPLIT}_C(X)$ recursively, as follows:
+We define $\operatorname{SPLIT}_{T, C}(X)$ recursively, as follows:
 
-- If $|X| = 0$, $\operatorname{SPLIT}_C(X) = \langle \rangle$
-- Otherwise, $\operatorname{SPLIT}_C(X) = \langle Y \rangle \mathbin{\|}
-  \operatorname{SPLIT}_C(Z)$ where
+- If $|X| = 0$, $\operatorname{SPLIT}_{T, C}(X) = \langle \rangle$
+- Otherwise, $\operatorname{SPLIT}_{T, C}(X) = \langle Y \rangle \mathbin{\|}
+  \operatorname{SPLIT}_{T, C}(Z)$ where
   - $i = I(X)$
   - $N = |X| - 1$
   - $Y = \langle X_0, \dots, X_i \rangle$
   - $Z = \langle X_{i+1}, \dots, X_N \rangle$
 
 # Tree Construction
 
-TODO
+For our purposes, a tree consists of a node (the "root") with an
+associated value (its "label") along with zero or more children, each
+of which is itself a tree. The children of each node are ordered. A
+"leaf" is a node with zero children; for convenience, a tree whose root
+is a leaf node will also be referred to as a "leaf".
+
+Let $C$ be a configuration, and let $T$ and $B$ be positive integers. We
+call $T$ the "threshold", as before, and $B$ the "step". Now for $X \in
+V_8$ we define $\operatorname{TREE}_{T, B, C} (X)$ recursively as
+follows:
+
+- If $I_{T, C} (X) = |X| - 1$, then $\operatorname{TREE}_{T, B, C}
+  (X)$ is a leaf labelled by $X$
+- Otherwise, let $n$ be the largest positive integer such that some
+  entry of $\operatorname{SPLIT}_{T + nB, C} (X)$, *other than the last
+  entry*, has length less than $S_{\text{max}}$, or zero if no such
+  positive integer exists. Then $\operatorname{TREE}_{T, B, C} (X)$ is a
+  node labelled by $X$ whose children in order are the subtrees
+  $\operatorname{TREE}_{T, B, C} (Y)$ for each entry $Y$ of
+  $\operatorname{SPLIT}_{T + nB, C} (X)$.
```

Comment:
I think this still has the problem with S_max: In the recursive case, simply shelling out to `SPLIT_{T + nB, C}` doesn't necessarliy give you a grouping of the chunks produced by `SPLIT_{T, C}`.

For example: if `SPLIT_{T, C}` makes it's first split at position `S_max - 4`, and its second at `S_max + 12`, assuming the first split didn't have any more bits set than necessary, `SPLIT_{T + nB, C}` will choose its first index as `S_max`, since it will hit that before the next viable split point. But that was *not* a split point with the smaller threshold. So you end up with different leaves, not just more of them.

Reply:
(Apologies for the delay in responding!) I guess I missed the point of your previous review, so thanks for spelling it out here. I agree that the algorithm has the property you describe, but I don't see why it's problematic -- does it interfere with the incremental recomputation procedure, or cause the constructed tree to differ from the bottom-up version, or something else? I am pretty sure that this step is at least well-defined, i.e. there exists a largest `n` satisfying the condition, because we can always pick an integer so large that the rolling hash never has enough trailing zeroes within our chunk and hence the only splits are caused by hitting the maximum size.

Reply:
It means it fails the last property in your list below:

> The labels of the leaf nodes in their natural order (inherited from the depth-first traversal of the tree) are exactly the entries of $\operatorname{SPLIT}_{T, C} (X)$.

...which I think is important: the idea is to arrange the chunks into a tree; it should not modify the split points of the chunks themselves.

(It does also differ from the bottom-up implementation, in that the bottom-implementation maintains that property).

Reply:
Oh, duh. Okay, it seems like we can fix that by waiting to apply the `S_max` cutoff: we go all the way down to `T + 0B` using just the hash-based splitting, and only then do we break down any remaining chunks bigger than `S_max`. That reflects how the bottom-up algorithm proceeds, only applying the cutoff if no "conventional" split intervenes first.

So the recursive definition will require a version of `SPLIT` that ignores `S_max` -- I'll edit that section accordingly. And since this approach has proved to be a bit tricky, it would be nice to include the bottom-up, iterative definition as well, though that probably belongs in a separate PR.

Reply:
I'm not sure that quite works either. Suppose the first natural split point is at position `S_max + N` where `N < S_min`. Because we hit `S_max` first, we'll split there, and then we will reject `S_max + N` as a split point because it results in a chunk that is too small. So you can't just split everything on natural boundaries and then add extra splits for `S_min`/`S_max` after the fact, since doing so can also require you to reject natural split points, not just add new split points.

I wonder if it is better to just use the bottom-up definition. Some arguments in favor:

- It lines up more closely to what implementations will do anyway.
  - We will want to document the bottom-up algorithm either way, since the correspondence is not entirely obvious.
- It's symmetric with what `SPLIT` does.

Especially if we start exploring things like min/max branching factors later, the symmetry with `SPLIT` is likely to be an asset.

I think before getting into the trenches the top-down definition seems more elegant & declarative, but I'm not sure what it buys us pragmatically, given how fiddly it has been to state it correctly; maybe just having the list of properties that hold is good enough.

Reply:
I'd thought `S_min` couldn't cause problems but you're quite right, it does. And the complexity of coalescing nodes at the end to repair the lower bound, and how that interacts with the fix for `S_max`, certainly seems like reason enough to abandon the top-down approach (certainly you'd never implement it this way, even in a purely functional language). I'll try writing up the bottom-up definition tonight.